In [1]:
import pandas as pd

my_dataset = pd.read_csv("gyro_acc_har_features.csv")



In [2]:
my_dataset.head()

,tBodyAcc-mean()-X,tBodyAcc-std()-X,tBodyAcc-min()-X,tBodyAcc-max()-X,tBodyAcc-energy()-X,tBodyAcc-entropy()-X,tBodyAcc-mean()-Y,tBodyAcc-std()-Y,tBodyAcc-min()-Y,tBodyAcc-max()-Y,...,fBodyGyroMag-std(),fBodyGyroMag-energy(),fBodyGyroMag-entropy(),fBodyGyroMag-meanFreq(),fBodyGyroJerkMag-mean(),fBodyGyroJerkMag-std(),fBodyGyroJerkMag-energy(),fBodyGyroJerkMag-entropy(),fBodyGyroJerkMag-meanFreq(),label
0,-0.049035,0.114672,-0.279797,0.405987,0.015554,2.686934,0.031145,0.137029,-0.389781,0.482584,...,20.742604,497.046709,0.496701,4.879113,320.145906,364.644965,235459.351647,1.201316,9.610271,STANDING
1,0.006539,0.157830,-0.351428,0.271580,0.024953,3.070939,0.022344,0.131549,-0.273999,0.407552,...,10.183367,124.894437,0.753526,5.280982,151.055706,168.856567,51330.366650,1.160908,9.601974,STANDING
2,0.095127,0.196963,-0.253708,0.675198,0.047844,2.663944,-0.007992,0.150100,-0.293755,0.503829,...,17.931015,393.398632,0.822879,4.277960,180.732863,185.868869,67211.604474,1.522016,9.739804,STANDING
3,0.085873,0.158424,-0.228166,0.679932,0.032472,2.687474,0.007704,0.120189,-0.329697,0.375013,...,26.449433,781.063946,0.425830,4.270597,249.688508,307.353871,156810.752914,1.140865,10.148589,STANDING
4,0.002873,0.153112,-0.404485,0.526157,0.023451,2.633197,0.003690,0.126452,-0.307763,0.340881,...,24.048815,641.426157,0.425830,4.497831,247.799944,338.385898,175909.828322,0.920849,10.438297,STANDING


In [3]:
my_dataset.describe()

,tBodyAcc-mean()-X,tBodyAcc-std()-X,tBodyAcc-min()-X,tBodyAcc-max()-X,tBodyAcc-energy()-X,tBodyAcc-entropy()-X,tBodyAcc-mean()-Y,tBodyAcc-std()-Y,tBodyAcc-min()-Y,tBodyAcc-max()-Y,...,fBodyGyroMag-mean(),fBodyGyroMag-std(),fBodyGyroMag-energy(),fBodyGyroMag-entropy(),fBodyGyroMag-meanFreq(),fBodyGyroJerkMag-mean(),fBodyGyroJerkMag-std(),fBodyGyroJerkMag-energy(),fBodyGyroJerkMag-entropy(),fBodyGyroJerkMag-meanFreq()
count,6.170000e+02,6.170000e+02,6.170000e+02,6.170000e+02,6.170000e+02,617.000000,617.000000,6.170000e+02,617.000000,6.170000e+02,...,617.000000,617.000000,617.000000,617.000000,617.000000,617.000000,617.000000,6.170000e+02,617.000000,6.170000e+02
mean,5.453458e-05,8.439842e-02,-2.236682e-01,2.247715e-01,1.341929e-02,2.776067,0.000416,9.033475e-02,-0.205172,2.878390e-01,...,2.768962,8.183021,137.895793,0.416974,3.720702,74.548984,101.176099,3.355077e+04,0.883935,9.590911e+00
std,1.944697e-02,7.699499e-02,2.232990e-01,2.389543e-01,2.064848e-02,0.297680,0.017570,1.107752e-01,0.252751,3.285817e-01,...,2.604531,7.522362,217.638535,0.195994,1.126052,80.495499,106.329280,6.133408e+04,0.281630,1.635613e+00
min,-1.204165e-01,2.129125e-07,-1.299157e+00,6.579159e-07,6.953612e-14,0.757136,-0.240845,5.535684e-07,-2.174850,1.728106e-07,...,0.105750,0.492638,0.272082,0.114676,0.000115,0.000002,0.000016,2.486050e-10,0.116115,8.296612e-10
25%,-3.948696e-03,1.190101e-02,-3.585779e-01,3.295223e-02,1.439552e-04,2.627797,-0.001463,5.791296e-03,-0.315334,1.613904e-02,...,0.395862,1.256015,1.720134,0.312211,3.044239,9.203063,11.841823,2.216990e+02,0.724590,9.087916e+00
50%,1.555746e-07,7.061364e-02,-1.608471e-01,1.637583e-01,5.235831e-03,2.846656,0.000060,4.592594e-02,-0.105089,1.432942e-01,...,2.309432,6.365432,45.571779,0.425830,3.557089,43.894357,57.524613,5.496246e+03,0.885196,9.869516e+00
75%,4.815555e-03,1.411635e-01,-3.049078e-02,3.516722e-01,2.043200e-02,2.973609,0.003286,1.370290e-01,-0.015227,5.019222e-01,...,4.499491,13.485003,203.578907,0.504128,4.420230,122.874280,169.579740,4.667303e+04,1.034889,1.046627e+01
max,1.365523e-01,4.463057e-01,-6.648181e-08,1.796449e+00,2.136889e-01,3.262262,0.077031,5.791504e-01,-0.000002,1.393172e+00,...,18.573167,39.200018,1623.544071,1.561467,8.058424,639.835009,513.152849,6.639583e+05,2.106231,1.181774e+01


In [4]:
new_dataset = pd.read_json(r"E:\src\neat-calculator\ml_monolith\data_preprocessing\merged_sensor_data_26012026_labeled.json")

new_dataset.describe()

,accelerometerX,accelerometerY,accelerometerZ,gyroscopeX,gyroscopeY,gyroscopeZ,timestamp,timestampNanos
count,31201.000000,31201.000000,31201.000000,31201.000000,31201.000000,31201.000000,31201,31201
mean,-3.017197,7.526292,-4.008101,0.005228,0.033317,-0.028379,2026-01-26 18:37:01.776114176,2012-05-08 10:14:08.399560704
min,-18.230000,-7.460000,-18.199999,-7.240000,-7.180000,-2.890000,2026-01-26 18:34:04.138000,2006-09-21 10:30:46.752000
25%,-3.770000,5.570000,-7.260000,-0.130000,-0.110000,-0.110000,2026-01-26 18:36:22.628999936,2011-02-10 08:11:04.043000064
50%,-2.880000,8.550000,-3.060000,0.000000,0.000000,0.000000,2026-01-26 18:37:09.268000,2012-08-03 03:35:09.336999936
75%,-2.450000,9.420000,-2.060000,0.120000,0.180000,0.060000,2026-01-26 18:38:06.988000,2014-06-02 04:51:08.534000128
max,19.670000,19.430000,15.880000,3.400000,3.320000,5.600000,2026-01-26 18:39:44.650000,2017-07-06 12:52:06.622000
std,1.499452,3.348278,3.144843,0.589227,0.756989,0.383560,NaN,NaN


In [10]:
import sys
import os
import importlib

# Add the parent directory (ml_monolith) to sys.path to access sibling folders
sys.path.append(os.path.abspath(os.path.join(os.getcwd(), '..')))

import data_pipeline.process_raw_data
importlib.reload(data_pipeline.process_raw_data)
from data_pipeline.process_raw_data import process_raw_sensor_data

new_dataset_processed = process_raw_sensor_data(raw_data_file_dir=r"E:\src\neat-calculator\data_collection\recordings\sensor_data\SensorRecording\26012026",
                                                kaggle_csv_path=r"E:\src\neat-calculator\ml_monolith\kaggle.csv",
                                                labels_csv_path=r"E:\src\neat-calculator\ml_monolith\data_preprocessing\labels26012026.csv")

new_dataset_processed.describe()

2026-02-07 10:44:13,327 - data_pipeline.process_raw_data - INFO - Starting processing of raw data in directory: E:\src\neat-calculator\data_collection\recordings\sensor_data\SensorRecording\26012026
2026-02-07 10:44:13,328 - data_pipeline.process_raw_data - INFO - Using labels from: E:\src\neat-calculator\ml_monolith\data_preprocessing\labels26012026.csv
2026-02-07 10:44:13,328 - data_pipeline.process_raw_data - INFO - Validating directory: E:\src\neat-calculator\data_collection\recordings\sensor_data\SensorRecording\26012026
2026-02-07 10:44:13,332 - data_pipeline.process_raw_data - INFO - Directory E:\src\neat-calculator\data_collection\recordings\sensor_data\SensorRecording\26012026 exists.
2026-02-07 10:44:13,333 - data_pipeline.process_raw_data - INFO - Path E:\src\neat-calculator\data_collection\recordings\sensor_data\SensorRecording\26012026 is a directory.
2026-02-07 10:44:13,335 - data_pipeline.process_raw_data - INFO - Directory E:\src\neat-calculator\data_collection\recordin

{'accelerometerX': 'mean', 'accelerometerY': 'mean', 'accelerometerZ': 'mean', 'gyroscopeX': 'mean', 'gyroscopeY': 'mean', 'gyroscopeZ': 'mean'}
{'accelerometerX': 'mean', 'accelerometerY': 'mean', 'accelerometerZ': 'mean', 'gyroscopeX': 'mean', 'gyroscopeY': 'mean', 'gyroscopeZ': 'mean', 'label': 'first', 'timestamp': 'first'}
                         dt_seconds  session_id
t                                              
2026-01-26 18:34:04.120        0.00           0
2026-01-26 18:34:04.140        0.02           0
2026-01-26 18:34:04.160        0.02           0
2026-01-26 18:34:04.180        0.02           0
2026-01-26 18:34:04.200        0.02           0
2026-01-26 18:34:04.220        0.02           0
2026-01-26 18:34:04.240        0.02           0
2026-01-26 18:34:04.260        0.02           0
2026-01-26 18:34:04.280        0.02           0
2026-01-26 18:34:04.300        0.02           0


2026-02-07 10:44:14,270 - data_pipeline.process_raw_data - INFO - Sliding windows created.
2026-02-07 10:44:14,270 - data_pipeline.process_raw_data - INFO - Columns after creating sliding windows: ['accelerometerX', 'accelerometerY', 'accelerometerZ', 'gyroscopeX', 'gyroscopeY', 'gyroscopeZ', 'label', 'timestamp', 'dt', 'dt_seconds', 'session_break', 'session_id']
2026-02-07 10:44:14,270 - data_pipeline.process_raw_data - INFO - Extracting features from sliding windows.


Number of windows: 258
Window shape: (128, 6)
Example labels: ['LAYING' 'SITTING' 'STANDING' 'WALKING' 'WALKING_DOWNSTAIRS'
 'WALKING_UPSTAIRS']


2026-02-07 10:44:18,190 - data_pipeline.process_raw_data - INFO - Shape of features: (258, 212)
2026-02-07 10:44:18,190 - data_pipeline.process_raw_data - INFO - Feature columns: Index(['tBodyAcc-mean()-X', 'tBodyAcc-std()-X', 'tBodyAcc-min()-X',
       'tBodyAcc-max()-X', 'tBodyAcc-energy()-X', 'tBodyAcc-entropy()-X',
       'tBodyAcc-mean()-Y', 'tBodyAcc-std()-Y', 'tBodyAcc-min()-Y',
       'tBodyAcc-max()-Y',
       ...
       'fBodyGyroMag-mean()', 'fBodyGyroMag-std()', 'fBodyGyroMag-energy()',
       'fBodyGyroMag-entropy()', 'fBodyGyroMag-meanFreq()',
       'fBodyGyroJerkMag-mean()', 'fBodyGyroJerkMag-std()',
       'fBodyGyroJerkMag-energy()', 'fBodyGyroJerkMag-entropy()',
       'fBodyGyroJerkMag-meanFreq()'],
      dtype='object', length=212)
2026-02-07 10:44:18,190 - data_pipeline.process_raw_data - INFO - Feature extraction completed.
2026-02-07 10:44:18,204 - data_pipeline.process_raw_data - INFO - Renaming features for clarity.
2026-02-07 10:44:18,207 - data_pipeline.proc

,tBodyAcc-mean()-X,tBodyAcc-std()-X,tBodyAcc-min()-X,tBodyAcc-max()-X,tBodyAcc-energy()-X,tBodyAcc-entropy()-X,tBodyAcc-mean()-Y,tBodyAcc-std()-Y,tBodyAcc-min()-Y,tBodyAcc-max()-Y,...,fBodyAccJerk-std()-Z,fBodyAccJerk-energy()-Z,fBodyAccJerk-entropy()-Z,fBodyAccJerk-meanFreq()-Z,fBodyAccMag-mean(),fBodyAccMag-std(),fBodyAccMag-energy(),fBodyAccMag-entropy(),fBodyAccMag-meanFreq(),timestamp
count,258.000000,258.000000,258.000000,258.000000,258.000000,258.000000,258.000000,258.000000,258.000000,258.000000,...,258.000000,258.000000,258.000000,258.000000,258.000000,258.000000,258.000000,258.000000,258.000000,2.580000e+02
mean,0.000328,0.112671,-0.300911,0.299098,0.022257,2.789732,0.000963,0.154298,-0.343812,0.445798,...,20.042847,3147.170282,2.922827,12.846831,1.298616,3.166825,20.494009,0.431165,5.700913,1.769453e+12
std,0.015898,0.096677,0.285278,0.309904,0.024271,0.356859,0.022978,0.142680,0.326755,0.419612,...,18.815013,3909.883139,0.426225,1.701535,1.132455,2.744181,21.746892,0.171688,0.902973,9.941792e+04
min,-0.066547,0.000888,-1.625518,0.002743,0.000001,1.156130,-0.119848,0.001964,-1.647336,0.001882,...,0.006266,0.000043,0.347266,3.046750,0.019709,0.052844,0.003231,0.114676,3.244478,1.769452e+12
25%,-0.005132,0.005964,-0.490335,0.015362,0.000036,2.667358,-0.004490,0.006473,-0.565529,0.016357,...,0.906234,3.186423,2.871322,12.459906,0.066865,0.141159,0.024090,0.312211,5.084254,1.769453e+12
50%,0.000081,0.124155,-0.285315,0.305964,0.016220,2.885922,0.000034,0.146843,-0.305254,0.415950,...,16.906913,1105.555294,2.992354,13.043789,1.250276,3.134775,11.957638,0.383447,5.792438,1.769453e+12
75%,0.004997,0.197964,-0.013285,0.486342,0.039428,3.024652,0.009095,0.278712,-0.018613,0.826393,...,35.768038,5263.059654,3.100152,13.547467,2.363616,5.824521,39.536068,0.496701,6.297586,1.769453e+12
max,0.069753,0.330397,-0.000277,2.357795,0.112824,3.244325,0.142490,0.441385,-0.003538,1.417123,...,79.418642,22199.200955,3.270733,15.207685,3.706019,7.796082,70.293750,1.401301,8.314648,1.769453e+12


In [13]:
new_dataset_processed.columns

Index(['tBodyAcc-mean()-X', 'tBodyAcc-std()-X', 'tBodyAcc-min()-X',
       'tBodyAcc-max()-X', 'tBodyAcc-energy()-X', 'tBodyAcc-entropy()-X',
       'tBodyAcc-mean()-Y', 'tBodyAcc-std()-Y', 'tBodyAcc-min()-Y',
       'tBodyAcc-max()-Y',
       ...
       'fBodyAccJerk-energy()-Z', 'fBodyAccJerk-entropy()-Z',
       'fBodyAccJerk-meanFreq()-Z', 'fBodyAccMag-mean()', 'fBodyAccMag-std()',
       'fBodyAccMag-energy()', 'fBodyAccMag-entropy()',
       'fBodyAccMag-meanFreq()', 'Activity', 'timestamp'],
      dtype='object', length=182)

In [15]:

from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
scaled_features = scaler.fit_transform(new_dataset_processed.drop(columns=["Activity"]))
scaled_df = pd.DataFrame(scaled_features, columns=new_dataset_processed.columns[:-1])
scaled_df.describe()

,tBodyAcc-mean()-X,tBodyAcc-std()-X,tBodyAcc-min()-X,tBodyAcc-max()-X,tBodyAcc-energy()-X,tBodyAcc-entropy()-X,tBodyAcc-mean()-Y,tBodyAcc-std()-Y,tBodyAcc-min()-Y,tBodyAcc-max()-Y,...,fBodyAccJerk-std()-Z,fBodyAccJerk-energy()-Z,fBodyAccJerk-entropy()-Z,fBodyAccJerk-meanFreq()-Z,fBodyAccMag-mean(),fBodyAccMag-std(),fBodyAccMag-energy(),fBodyAccMag-entropy(),fBodyAccMag-meanFreq(),Activity
count,258.000000,258.000000,258.000000,258.000000,258.000000,258.000000,258.000000,258.000000,258.000000,258.000000,...,258.000000,258.000000,258.000000,258.000000,258.000000,258.000000,258.000000,258.000000,258.000000,258.000000
mean,0.490643,0.339240,0.815022,0.125838,0.197261,0.782303,0.460517,0.346671,0.792995,0.313668,...,0.252311,0.141770,0.880995,0.805866,0.346934,0.402155,0.291516,0.245984,0.484488,0.501776
std,0.116638,0.293398,0.175529,0.131591,0.215123,0.170893,0.087588,0.324700,0.198780,0.296495,...,0.236928,0.176127,0.145794,0.139918,0.307205,0.354397,0.309386,0.133441,0.178095,0.294205
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,0.450584,0.015405,0.698471,0.005358,0.000306,0.723700,0.439733,0.010262,0.658114,0.010228,...,0.011333,0.000144,0.863378,0.774049,0.012792,0.011405,0.000297,0.153530,0.362863,0.243373
50%,0.488830,0.374093,0.824618,0.128753,0.143753,0.828367,0.456977,0.329704,0.816452,0.292577,...,0.212821,0.049802,0.904778,0.822062,0.333821,0.398016,0.170071,0.208896,0.502539,0.513246
75%,0.524902,0.598090,0.991997,0.205345,0.349454,0.894802,0.491516,0.629803,0.990829,0.582594,...,0.450330,0.237083,0.941651,0.863479,0.635841,0.745383,0.562421,0.296921,0.602171,0.756584
max,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,...,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000


In [17]:
# rename angle(tBodyAccJerkMean),gravityMean) in kaggle dataset to match my_dataset

kaggle_dataset = pd.read_csv(r"E:\src\neat-calculator\ml_monolith\kaggle.csv")
kaggle_dataset = kaggle_dataset.rename(columns={"angle(tBodyAccJerkMean),gravityMean)": "angle(tBodyAccJerkMean,gravityMean)"})
kaggle_dataset = kaggle_dataset.rename(columns={"angle(tBodyAccMean,gravity)": "angle(tBodyAccMean,gravityMean)"})
kaggle_dataset = kaggle_dataset.rename(columns={"fBodyBodyAccJerkMag-mean()": "fBodyAccJerkMag-mean()"})

# rename all columns containing "BodyBody" to "Body" in kaggle_dataset
kaggle_dataset.columns = kaggle_dataset.columns.str.replace("BodyBody", "Body")

kaggle_dataset.describe()

,tBodyAcc-mean()-X,tBodyAcc-mean()-Y,tBodyAcc-mean()-Z,tBodyAcc-std()-X,tBodyAcc-std()-Y,tBodyAcc-std()-Z,tBodyAcc-mad()-X,tBodyAcc-mad()-Y,tBodyAcc-mad()-Z,tBodyAcc-max()-X,...,fBodyGyroJerkMag-skewness(),fBodyGyroJerkMag-kurtosis(),"angle(tBodyAccMean,gravityMean)","angle(tBodyAccJerkMean,gravityMean)","angle(tBodyGyroMean,gravityMean)","angle(tBodyGyroJerkMean,gravityMean)","angle(X,gravityMean)","angle(Y,gravityMean)","angle(Z,gravityMean)",subject
count,7352.000000,7352.000000,7352.000000,7352.000000,7352.000000,7352.000000,7352.000000,7352.000000,7352.000000,7352.000000,...,7352.000000,7352.000000,7352.000000,7352.000000,7352.000000,7352.000000,7352.000000,7352.000000,7352.000000,7352.000000
mean,0.274488,-0.017695,-0.109141,-0.605438,-0.510938,-0.604754,-0.630512,-0.526907,-0.606150,-0.468604,...,-0.307009,-0.625294,0.008684,0.002186,0.008726,-0.005981,-0.489547,0.058593,-0.056515,17.413085
std,0.070261,0.040811,0.056635,0.448734,0.502645,0.418687,0.424073,0.485942,0.414122,0.544547,...,0.321011,0.307584,0.336787,0.448306,0.608303,0.477975,0.511807,0.297480,0.279122,8.975143
min,-1.000000,-1.000000,-1.000000,-1.000000,-0.999873,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,...,-0.995357,-0.999765,-0.976580,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,1.000000
25%,0.262975,-0.024863,-0.120993,-0.992754,-0.978129,-0.980233,-0.993591,-0.978162,-0.980251,-0.936219,...,-0.542602,-0.845573,-0.121527,-0.289549,-0.482273,-0.376341,-0.812065,-0.017885,-0.143414,8.000000
50%,0.277193,-0.017219,-0.108676,-0.946196,-0.851897,-0.859365,-0.950709,-0.857328,-0.857143,-0.881637,...,-0.343685,-0.711692,0.009509,0.008943,0.008735,-0.000368,-0.709417,0.182071,0.003181,19.000000
75%,0.288461,-0.010783,-0.097794,-0.242813,-0.034231,-0.262415,-0.292680,-0.066701,-0.265671,-0.017129,...,-0.126979,-0.503878,0.150865,0.292861,0.506187,0.359368,-0.509079,0.248353,0.107659,26.000000
max,1.000000,1.000000,1.000000,1.000000,0.916238,1.000000,1.000000,0.967664,1.000000,1.000000,...,0.989538,0.956845,1.000000,1.000000,0.998702,0.996078,1.000000,0.478157,1.000000,30.000000


In [5]:
my_dataset = my_dataset.rename(columns={"label": "Activity"})
my_dataset = my_dataset.rename(columns={"subject": "Subject"})

In [6]:
# create a new dataset containing only the columns that are present in both datasets

combined_columns = my_dataset.columns.intersection(kaggle_dataset.columns)
my_dataset = my_dataset.reindex(columns=combined_columns)
kaggle_dataset = kaggle_dataset.reindex(columns=combined_columns)


In [7]:
my_dataset.describe()

,tBodyAcc-mean()-X,tBodyAcc-std()-X,tBodyAcc-min()-X,tBodyAcc-max()-X,tBodyAcc-energy()-X,tBodyAcc-entropy()-X,tBodyAcc-mean()-Y,tBodyAcc-std()-Y,tBodyAcc-min()-Y,tBodyAcc-max()-Y,...,fBodyGyroMag-mean(),fBodyGyroMag-std(),fBodyGyroMag-energy(),fBodyGyroMag-entropy(),fBodyGyroMag-meanFreq(),fBodyGyroJerkMag-mean(),fBodyGyroJerkMag-std(),fBodyGyroJerkMag-energy(),fBodyGyroJerkMag-entropy(),fBodyGyroJerkMag-meanFreq()
count,6.170000e+02,6.170000e+02,6.170000e+02,6.170000e+02,6.170000e+02,617.000000,617.000000,6.170000e+02,617.000000,6.170000e+02,...,617.000000,617.000000,617.000000,617.000000,617.000000,617.000000,617.000000,6.170000e+02,617.000000,6.170000e+02
mean,5.453458e-05,8.439842e-02,-2.236682e-01,2.247715e-01,1.341929e-02,2.776067,0.000416,9.033475e-02,-0.205172,2.878390e-01,...,2.768962,8.183021,137.895793,0.416974,3.720702,74.548984,101.176099,3.355077e+04,0.883935,9.590911e+00
std,1.944697e-02,7.699499e-02,2.232990e-01,2.389543e-01,2.064848e-02,0.297680,0.017570,1.107752e-01,0.252751,3.285817e-01,...,2.604531,7.522362,217.638535,0.195994,1.126052,80.495499,106.329280,6.133408e+04,0.281630,1.635613e+00
min,-1.204165e-01,2.129125e-07,-1.299157e+00,6.579159e-07,6.953612e-14,0.757136,-0.240845,5.535684e-07,-2.174850,1.728106e-07,...,0.105750,0.492638,0.272082,0.114676,0.000115,0.000002,0.000016,2.486050e-10,0.116115,8.296612e-10
25%,-3.948696e-03,1.190101e-02,-3.585779e-01,3.295223e-02,1.439552e-04,2.627797,-0.001463,5.791296e-03,-0.315334,1.613904e-02,...,0.395862,1.256015,1.720134,0.312211,3.044239,9.203063,11.841823,2.216990e+02,0.724590,9.087916e+00
50%,1.555746e-07,7.061364e-02,-1.608471e-01,1.637583e-01,5.235831e-03,2.846656,0.000060,4.592594e-02,-0.105089,1.432942e-01,...,2.309432,6.365432,45.571779,0.425830,3.557089,43.894357,57.524613,5.496246e+03,0.885196,9.869516e+00
75%,4.815555e-03,1.411635e-01,-3.049078e-02,3.516722e-01,2.043200e-02,2.973609,0.003286,1.370290e-01,-0.015227,5.019222e-01,...,4.499491,13.485003,203.578907,0.504128,4.420230,122.874280,169.579740,4.667303e+04,1.034889,1.046627e+01
max,1.365523e-01,4.463057e-01,-6.648181e-08,1.796449e+00,2.136889e-01,3.262262,0.077031,5.791504e-01,-0.000002,1.393172e+00,...,18.573167,39.200018,1623.544071,1.561467,8.058424,639.835009,513.152849,6.639583e+05,2.106231,1.181774e+01


In [7]:
my_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 617 entries, 0 to 616
Columns: 198 entries, tBodyAcc-mean()-X to Activity
dtypes: float64(197), object(1)
memory usage: 954.5+ KB


In [9]:
# my dataset feature count
print(f"My Dataset Feature Count: {my_dataset.shape[1]}")
# kaggle dataset feature count
print(f"Kaggle Dataset Feature Count: {kaggle_dataset.shape[1]}")

print("Columns missing in my data:")
print(set(kaggle_dataset.columns) - set(my_dataset.columns))

print("Extra columns in my data:")
print(set(my_dataset.columns) - set(kaggle_dataset.columns))

print("Columns in both datasets:")
print(set(my_dataset.columns).intersection(set(kaggle_dataset.columns)))

My Dataset Feature Count: 198
Kaggle Dataset Feature Count: 198
Columns missing in my data:
set()
Extra columns in my data:
set()
Columns in both datasets:
{'tBodyGyro-entropy()-X', 'tBodyGyro-energy()-Z', 'tBodyGyroJerk-entropy()-Z', 'tBodyGyroJerk-std()-X', 'fBodyGyroJerkMag-entropy()', 'tBodyGyroJerk-max()-X', 'tBodyGyroJerk-entropy()-X', 'tBodyGyro-max()-X', 'fBodyAcc-mean()-Y', 'fBodyAccJerk-std()-Z', 'tBodyGyro-mean()-Z', 'tBodyAcc-entropy()-Z', 'tGravityAcc-std()-X', 'tGravityAccMag-mean()', 'tGravityAcc-min()-X', 'tBodyAccJerkMag-min()', 'fBodyAccJerk-energy()-X', 'tBodyGyroJerkMag-entropy()', 'tBodyGyroMag-energy()', 'tBodyGyroJerk-min()-Z', 'tGravityAcc-entropy()-X', 'tBodyGyro-min()-X', 'fBodyAcc-entropy()-Y', 'fBodyAccJerk-mean()-Y', 'tBodyGyroJerk-mean()-X', 'fBodyGyro-mean()-Z', 'fBodyGyroJerkMag-mean()', 'fBodyGyro-entropy()-Y', 'tGravityAccMag-min()', 'tBodyAccMag-entropy()', 'tBodyGyroJerk-max()-Z', 'tBodyGyroMag-std()', 'fBodyAcc-energy()-X', 'fBodyAccJerk-mean()-X', 

In [9]:
kaggle_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7352 entries, 0 to 7351
Columns: 198 entries, tBodyAcc-mean()-X to Activity
dtypes: float64(197), object(1)
memory usage: 11.1+ MB


at this point we have 616 records in my dataset and 7352 records in the kaggle dataset. now it's time to prepare some baseline models for both 

In [21]:
# My dataset RANDOM FOREST
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix

X = my_dataset.drop("Activity", axis=1)
y = my_dataset["Activity"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
rf = RandomForestClassifier(n_estimators=100, random_state=42)
rf.fit(X_train, y_train)
y_pred = rf.predict(X_test)
print("My Dataset Random Forest Classification Report:")
print(classification_report(y_test, y_pred))
print("My Dataset Random Forest Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))


My Dataset Random Forest Classification Report:
                    precision    recall  f1-score   support

           SITTING       1.00      0.50      0.67         6
          STANDING       0.89      0.98      0.93        87
           WALKING       0.70      0.61      0.65        23
WALKING_DOWNSTAIRS       0.83      0.62      0.71         8

          accuracy                           0.86       124
         macro avg       0.86      0.68      0.74       124
      weighted avg       0.86      0.86      0.85       124

My Dataset Random Forest Confusion Matrix:
[[ 3  2  1  0]
 [ 0 85  2  0]
 [ 0  8 14  1]
 [ 0  0  3  5]]


My dataset isn't very large and the classes aren't distributed equally - there is very little samples of Walking downstairs and a lot of examples of standing and walking. Again, not many examples of Sitting.

still the baseline model has quite okayish accuracy.


In [11]:
# KAGGGLE dataset RANDOM FOREST
X_kaggle = kaggle_dataset.drop("Activity", axis=1)
y_kaggle = kaggle_dataset["Activity"]
Xk_train, Xk_test, yk_train, yk_test = train_test_split(X_kaggle, y_kaggle, test_size=0.2, random_state=42)
rf_kaggle = RandomForestClassifier(n_estimators=100, random_state=42)
rf_kaggle.fit(Xk_train, yk_train)
yk_pred = rf_kaggle.predict(Xk_test)
print("Kaggle Dataset Random Forest Classification Report:")
print(classification_report(yk_test, yk_pred))
print("Kaggle Dataset Random Forest Confusion Matrix:")
print(confusion_matrix(yk_test, yk_pred))


Kaggle Dataset Random Forest Classification Report:
                    precision    recall  f1-score   support

            LAYING       1.00      1.00      1.00       280
           SITTING       0.96      0.96      0.96       262
          STANDING       0.96      0.96      0.96       276
           WALKING       1.00      0.99      0.99       247
WALKING_DOWNSTAIRS       0.99      0.99      0.99       206
  WALKING_UPSTAIRS       0.99      0.99      0.99       200

          accuracy                           0.98      1471
         macro avg       0.98      0.98      0.98      1471
      weighted avg       0.98      0.98      0.98      1471

Kaggle Dataset Random Forest Confusion Matrix:
[[280   0   0   0   0   0]
 [  0 252  10   0   0   0]
 [  0  10 266   0   0   0]
 [  0   0   0 245   1   1]
 [  0   0   0   1 203   2]
 [  0   0   0   0   2 198]]


Kaggle dataset performs much better because it has much more data and it has better quality. NOw i will check if my data will improve or destabilize the model. Later I will try to predict kaggle data on my dataset and my dataset on kaggle data

In [12]:
combined_datasets = pd.concat([my_dataset, kaggle_dataset], ignore_index=True)

X_combined = combined_datasets.drop("Activity", axis=1)
y_combined = combined_datasets["Activity"]
Xc_train, Xc_test, yc_train, yc_test = train_test_split(X_combined, y_combined, test_size=0.2, random_state=42)
rf_combined = RandomForestClassifier(n_estimators=100, random_state=42)
rf_combined.fit(Xc_train, yc_train)
yc_pred = rf_combined.predict(Xc_test)
print("Combined Dataset Random Forest Classification Report:")
print(classification_report(yc_test, yc_pred))
print("Combined Dataset Random Forest Confusion Matrix:")
print(confusion_matrix(yc_test, yc_pred))

Combined Dataset Random Forest Classification Report:
                    precision    recall  f1-score   support

            LAYING       1.00      1.00      1.00       266
           SITTING       0.99      0.96      0.98       254
          STANDING       0.95      0.98      0.97       375
           WALKING       0.96      0.96      0.96       294
WALKING_DOWNSTAIRS       0.99      0.97      0.98       202
  WALKING_UPSTAIRS       0.98      1.00      0.99       203

          accuracy                           0.98      1594
         macro avg       0.98      0.98      0.98      1594
      weighted avg       0.98      0.98      0.98      1594

Combined Dataset Random Forest Confusion Matrix:
[[266   0   0   0   0   0]
 [  0 244   7   2   0   1]
 [  0   2 368   5   0   0]
 [  0   0  12 282   0   0]
 [  0   0   0   4 195   3]
 [  0   0   0   0   1 202]]


In [20]:
# because my dataset and kaggle have differnet labels - i am going to remove from the kaggle datasets all rows with labels- LAYING, WALKING_UPSTAIRS

kaggle_dataset_filtered =  kaggle_dataset[~kaggle_dataset['Activity'].isin(['LAYING', 'WALKING_UPSTAIRS'])]




In [22]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_my_test = my_dataset.drop("Activity", axis=1)
y_my_test = my_dataset["Activity"]
X_my_test_scaled = scaler.fit_transform(X_my_test)
X_kaggle_train = kaggle_dataset_filtered.drop("Activity", axis=1)
y_kaggle_train = kaggle_dataset_filtered["Activity"]
X_kaggle_train_scaled = scaler.fit_transform(X_kaggle_train)


rf_kaggle_final = RandomForestClassifier(n_estimators=100, random_state=42)
rf_kaggle_final.fit(X_kaggle_train_scaled, y_kaggle_train)
y_my_test_pred = rf_kaggle_final.predict(X_my_test_scaled)
print("Kaggle Trained Model on My Test Data - Classification Report:")
print(classification_report(y_my_test, y_my_test_pred))
print("Kaggle Trained Model on My Test Data - Confusion Matrix:")
print(confusion_matrix(y_my_test, y_my_test_pred))

Kaggle Trained Model on My Test Data - Classification Report:
                    precision    recall  f1-score   support

           SITTING       0.05      0.21      0.08        28
          STANDING       0.93      0.44      0.60       423
           WALKING       0.45      0.52      0.48       132
WALKING_DOWNSTAIRS       0.18      0.79      0.30        34

          accuracy                           0.47       617
         macro avg       0.40      0.49      0.36       617
      weighted avg       0.75      0.47      0.53       617

Kaggle Trained Model on My Test Data - Confusion Matrix:
[[  6   8   1  13]
 [108 186  75  54]
 [  4   5  68  55]
 [  0   0   7  27]]


## Results After Fixes

**Major improvement from 4% to 47% accuracy!** ✓

The feature engineering fixes (entropy, gravity normalization) helped significantly:
- Before: Model predicted everything as WALKING_DOWNSTAIRS
- After: Model now recognizes all 4 activities with varying success

**Per-class performance:**
- STANDING: 93% precision, 44% recall - Good at identifying standing when it predicts it
- WALKING: 45% precision, 52% recall - Moderate performance  
- WALKING_DOWNSTAIRS: 79% recall - Good at catching most instances
- SITTING: 21% recall - Still struggles (confuses with STANDING)

The remaining gap is due to:
1. Different recording conditions (phone positions, movement patterns)
2. 34 features still have extreme values (jerk/frequency domain)
3. Limited training data in your dataset (617 samples vs 4872 Kaggle samples)

In [23]:
# Test reverse: Train on MY data, test on Kaggle (sample)
from sklearn.preprocessing import StandardScaler

# Sample from Kaggle for testing
kaggle_test_sample = kaggle_dataset_filtered.sample(n=100, random_state=42)

X_my_train = my_dataset.drop("Activity", axis=1)
y_my_train = my_dataset["Activity"]

X_kaggle_test = kaggle_test_sample.drop("Activity", axis=1)
y_kaggle_test = kaggle_test_sample["Activity"]

# Fit scaler on MY data, transform both
scaler_reverse = StandardScaler()
X_my_train_scaled = scaler_reverse.fit_transform(X_my_train)
X_kaggle_test_scaled = scaler_reverse.transform(X_kaggle_test)

# Train on my data, test on Kaggle
rf_reverse = RandomForestClassifier(n_estimators=100, random_state=42)
rf_reverse.fit(X_my_train_scaled, y_my_train)
y_kaggle_pred = rf_reverse.predict(X_kaggle_test_scaled)

print("=== MY Model Tested on Kaggle Data ===")
print(classification_report(y_kaggle_test, y_kaggle_pred))
print("\nConfusion Matrix:")
print(confusion_matrix(y_kaggle_test, y_kaggle_pred))

=== MY Model Tested on Kaggle Data ===
                    precision    recall  f1-score   support

           SITTING       0.24      1.00      0.39        24
          STANDING       0.00      0.00      0.00        24
           WALKING       0.00      0.00      0.00        27
WALKING_DOWNSTAIRS       0.00      0.00      0.00        25

          accuracy                           0.24       100
         macro avg       0.06      0.25      0.10       100
      weighted avg       0.06      0.24      0.09       100


Confusion Matrix:
[[24  0  0  0]
 [24  0  0  0]
 [27  0  0  0]
 [25  0  0  0]]


e:\src\neat-calculator\data_preprocessing\.venv\lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
e:\src\neat-calculator\data_preprocessing\.venv\lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
e:\src\neat-calculator\data_preprocessing\.venv\lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{me

In [26]:
# Combined dataset with PROPER train/test split

combined_datasets = pd.concat([my_dataset, kaggle_dataset], ignore_index=True)
combined_datasets.to_csv("E:\\src\\neat-calculator\\model_training\\combined_dataset.csv", index=False)


=== Combined Dataset with Proper Split ===
                    precision    recall  f1-score   support

            LAYING       1.00      1.00      1.00       281
           SITTING       0.94      0.96      0.95       263
          STANDING       0.97      0.96      0.96       359
           WALKING       0.97      0.99      0.98       272
WALKING_DOWNSTAIRS       0.99      0.97      0.98       204
  WALKING_UPSTAIRS       0.99      1.00      0.99       215

          accuracy                           0.98      1594
         macro avg       0.98      0.98      0.98      1594
      weighted avg       0.98      0.98      0.98      1594


Confusion Matrix:
[[281   0   0   0   0   0]
 [  0 253   9   1   0   0]
 [  0  15 343   1   0   0]
 [  0   0   3 268   0   1]
 [  0   0   0   5 198   1]
 [  0   0   0   0   1 214]]

Train size: 6375, Test size: 1594


## Final Summary

### Cross-Dataset Performance:
- **Kaggle → Your Data: 47%** ✓ (Major improvement from 4%)
- **Your Data → Kaggle: 24%** (Limited by small training set)
- **Within Your Data: 86%** ✓ (Good performance)
- **Within Kaggle: ~95%** ✓ (Excellent performance)
- **Combined Dataset: 98%** ✓ (Best performance)

### Impact of Fixes:
1. ✓ Fixed entropy calculations (eliminated extreme negative values)
2. ✓ Normalized accelerometer to gravity units
3. ✓ Improved from 4% to 47% cross-dataset accuracy

### Remaining Challenges:
- Different recording conditions (phone orientation, movement style)
- 34 features still have different scales (jerk/frequency domain)
- Limited training data (617 vs 4872 samples)

### Recommendations:
1. **Use the combined dataset** (98% accuracy) for production
2. **Collect more data** from your recording setup to improve generalization
3. **Consider per-subject normalization** if tracking specific users